torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -494,6 +496,256 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+// Create a TensorView before the original tensor. A common use case is to read
+// original tensor into shared memory or local registers.
+// Analogous to TVM Cache_Write
+TensorView* TensorView::cache_before() {
+  FusionGuard fg(this->fusion());
+
+  Expr* origin_expr = this->fusion()->origin(this);
+  TORCH_CHECK(
+      origin_expr != nullptr,
+      "Error adding cache_before ",
+      this,
+      " its origin is a nullptr and we restrict using cache_before on an input.");
+
+  // Create Producer Domain
+  auto dom = domain()->noReductions();
+  std::vector<IterDomain*> out_domain(dom.size(), nullptr);
+  for (size_t i = 0; i < dom.size(); ++i) {
+    if (dom[i]->isBroadcast()) {
+      out_domain[i] = new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true);
+    } else {
+      out_domain[i] = new IterDomain(dom[i]->start(), dom[i]->extent());
+    }
+  }
+
+  // This domain will be the consumer, so create the producer
+  TensorView* producer =
+      new TensorView(new TensorDomain(out_domain), this->getDataType().value());
+
+  // Set domain of consumer
+  TensorView* consumer = this;
+
+  // Setup dependency chain, inserting producer - Cache_Read - before this op.
+  // Before: Prev Op -> This Op
+  // After: Prev Op -> New Cache_Read Op -> This Op
+
+  // Expr* producer_origin =
+  if (origin_expr != nullptr)
+    this->createExprConsumer(origin_expr, producer);
+
+  // Expr* producer_uses =
+  new UnaryOp(UnaryOpType::Set, consumer, producer);
+
+  // ComputeAt - Before
+  if(this->hasComputeAt())
+    producer->computeAt(consumer, getThisComputeAtAxis());
```

Comment:
Once the above is fixed this should be fine.

You also need to run through tensors and find if anything was set to computeAt into this tensor, and redo it computeAt-ed into the new tensor.

I'm uncertain exactly how the computeAt position argument matches up, please think through it carefully, as computeAt axes can change based on producer consumer relationships and if there are reductions/broadcasts within the computeAt axis.